File: src/shared/pnHostingPolicy.ts
/** LXMF local PN hosting / peering policy (mirrors sidecar `PnHostingPolicy`). */

export interface PnHostingPolicy {
  peering_cost: number;
  max_peering_cost: number;
  autopeer: boolean;
  autopeer_maxdepth: number;
  max_peers: number;
  propagation_stamp_cost: number;
  propagation_stamp_flex: number;
  message_storage_limit_mb: number;
  propagation_limit_kb: number;
  sync_limit_kb: number;
  delivery_limit_kb: number;
  from_static_only: boolean;
  auth_required: boolean;
  enforce_stamps: boolean;
  enforce_ratchets: boolean;
  static_peers: string[];
  node_name: string | null;
  pn_announce_interval_sec: number;
  announce_at_start: boolean;
}

export const DEFAULT_PN_HOSTING_POLICY: PnHostingPolicy = {
  peering_cost: 18,
  max_peering_cost: 26,
  autopeer: true,
  autopeer_maxdepth: 4,
  max_peers: 20,
  propagation_stamp_cost: 16,
  propagation_stamp_flex: 3,
  message_storage_limit_mb: 256,
  propagation_limit_kb: 256,
  sync_limit_kb: 10_240,
  delivery_limit_kb: 1000,
  from_static_only: false,
  auth_required: false,
  enforce_stamps: false,
  enforce_ratchets: false,
  static_peers: [],
  node_name: null,
  pn_announce_interval_sec: 360,
  announce_at_start: true,
};

/** Mirrors Rust `MAX_*` caps in `pn_hosting_policy.rs`. */
const MAX_AUTOPEER_MAXDEPTH = 64;
const MAX_MAX_PEERS = 256;
const MAX_STORAGE_MB = 10_240;
const MAX_LIMIT_KB = 102_400;
const MAX_PN_ANNOUNCE_INTERVAL_SEC = 86_400;
const MAX_NODE_NAME_CHARS = 128;
const MAX_STATIC_PEERS = 256;

const NUMERIC_POLICY_KEYS = [
  'peering_cost',
  'max_peering_cost',
  'autopeer_maxdepth',
  'max_peers',
  'propagation_stamp_cost',
  'propagation_stamp_flex',
  'message_storage_limit_mb',
  'propagation_limit_kb',
  'sync_limit_kb',
  'delivery_limit_kb',
  'pn_announce_interval_sec',
] as const;

/** Sidecar serializes these as `u8` — reject values outside 0..=255 before IPC. */
const U8_POLICY_KEYS = [
  'peering_cost',
  'max_peering_cost',
  'propagation_stamp_cost',
  'propagation_stamp_flex',
] as const;

export type SanitizePnHostingPolicyResult =
  { ok: true; policy: PnHostingPolicy } | { ok: false; error: string };

function validateStaticPeerHash(hash: string): string | null {
  const trimmed = hash.trim().toLowerCase();
  if (trimmed.length !== 32 || !/^[0-9a-f]{32}$/.test(trimmed)) {
    return `static_peer_invalid:${trimmed}`;
  }
  return null;
}

/** Map validation / sidecar codes to i18n keys under `networkPanel.reticulumPnHosting.error.*`. */
export function mapPnHostingPolicyError(code: string | null | undefined): string {
  if (!code) return 'networkPanel.reticulumPnHosting.saveFailed';
  if (code.startsWith('static_peer_invalid')) {
    return 'networkPanel.reticulumPnHosting.error.staticPeerInvalid';
  }
  const known: Record<string, string> = {
    peering_cost_exceeds_max: 'networkPanel.reticulumPnHosting.error.peeringCostExceedsMax',
    stamp_flex_exceeds_cost: 'networkPanel.reticulumPnHosting.error.stampFlexExceedsCost',
    autopeer_maxdepth_out_of_range: 'networkPanel.reticulumPnHosting.error.autopeerMaxdepth',
    max_peers_out_of_range: 'networkPanel.reticulumPnHosting.error.maxPeers',
    message_storage_limit_out_of_range: 'networkPanel.reticulumPnHosting.error.storageMb',
    propagation_limit_out_of_range: 'networkPanel.reticulumPnHosting.error.propagationLimit',
    sync_limit_out_of_range: 'networkPanel.reticulumPnHosting.error.syncLimit',
    delivery_limit_out_of_range: 'networkPanel.reticulumPnHosting.error.deliveryLimit',
    pn_announce_interval_out_of_range: 'networkPanel.reticulumPnHosting.error.announceInterval',
    static_peers_too_many: 'networkPanel.reticulumPnHosting.error.staticPeersTooMany',
    node_name_invalid: 'networkPanel.reticulumPnHosting.error.nodeNameInvalid',
    node_name_too_long: 'networkPanel.reticulumPnHosting.error.nodeNameTooLong',
    non_finite_number: 'networkPanel.reticulumPnHosting.error.nonFiniteNumber',
  };
  return known[code] ?? 'networkPanel.reticulumPnHosting.saveFailed';
}

function validateStaticPeerList(peers: readonly string[]): string | null {
  if (peers.length > MAX_STATIC_PEERS) {
    return 'static_peers_too_many';
  }
  for (const peer of peers) {
    const peerErr = validateStaticPeerHash(peer);
    if (peerErr) return peerErr;
  }
  return null;
}

function validatePnHostingNodeName(nodeName: string | null): string | null {
  if (nodeName == null) return null;
  const trimmed = nodeName.trim();
  let scalarCount = 0;
  for (let i = 0; i < trimmed.length;) {
    const cp = trimmed.codePointAt(i);
    if (cp === undefined) break;
    if (cp < 32 || cp === 127) {
      return 'node_name_invalid';
    }
    scalarCount += 1;
    i += cp > 0xffff ? 2 : 1;
  }
  if (scalarCount > MAX_NODE_NAME_CHARS) {
    return 'node_name_too_long';
  }
  return null;
}

function validatePnHostingNumericRanges(policy: PnHostingPolicy): string | null {
  for (const key of NUMERIC_POLICY_KEYS) {
    const n = policy[key];
    if (!Number.isFinite(n) || !Number.isInteger(n) || n < 0) {
      return 'non_finite_number';
    }
  }
  for (const key of U8_POLICY_KEYS) {
    if (policy[key] > 255) {
      return 'non_finite_number';
    }
  }
  if (policy.peering_cost > policy.max_peering_cost) {
    return 'peering_cost_exceeds_max';
  }
  if (policy.propagation_stamp_flex > policy.propagation_stamp_cost) {
    return 'stamp_flex_exceeds_cost';
  }
  if (policy.autopeer_maxdepth > MAX_AUTOPEER_MAXDEPTH) {
    return 'autopeer_maxdepth_out_of_range';
  }
  if (policy.max_peers === 0 || policy.max_peers > MAX_MAX_PEERS) {
    return 'max_peers_out_of_range';
  }
  if (policy.message_storage_limit_mb === 0 || policy.message_storage_limit_mb > MAX_STORAGE_MB) {
    return 'message_storage_limit_out_of_range';
  }
  if (policy.propagation_limit_kb === 0 || policy.propagation_limit_kb > MAX_LIMIT_KB) {
    return 'propagation_limit_out_of_range';
  }
  if (policy.sync_limit_kb === 0 || policy.sync_limit_kb > MAX_LIMIT_KB) {
    return 'sync_limit_out_of_range';
  }
  if (policy.delivery_limit_kb === 0 || policy.delivery_limit_kb > MAX_LIMIT_KB) {
    return 'delivery_limit_out_of_range';
  }
  if (policy.pn_announce_interval_sec > MAX_PN_ANNOUNCE_INTERVAL_SEC) {
    return 'pn_announce_interval_out_of_range';
  }
  return null;
}

/** Semantic checks matching sidecar `PnHostingPolicy::validate`. */
export function validatePnHostingPolicy(policy: PnHostingPolicy): string | null {
  return (
    validatePnHostingNumericRanges(policy) ??
    validateStaticPeerList(policy.static_peers) ??
    validatePnHostingNodeName(policy.node_name)
  );
}

/**
 * Normalize peers/name then validate — mirrors sidecar `PnHostingPolicy::sanitized`.
 */
export function sanitizePnHostingPolicy(policy: PnHostingPolicy): SanitizePnHostingPolicyResult {
  const staticPeers = policy.static_peers
    .map((s) => s.trim().toLowerCase())
    .filter((s) => s.length > 0);
  let nodeName: string | null = null;
  if (policy.node_name != null) {
    const trimmed = policy.node_name.trim();
    nodeName = trimmed.length > 0 ? trimmed : null;
  }
  const next: PnHostingPolicy = {
    ...policy,
    static_peers: staticPeers,
    node_name: nodeName,
  };
  const error = validatePnHostingPolicy(next);
  if (error) return { ok: false, error };
  return { ok: true, policy: next };
}

export function parsePnHostingPolicy(raw: unknown): PnHostingPolicy {
  if (!raw || typeof raw !== 'object') {
    return { ...DEFAULT_PN_HOSTING_POLICY };
  }
  const o = raw as Record<string, unknown>;
  const num = (key: keyof PnHostingPolicy, fallback: number): number => {
    const v = o[key];
    return typeof v === 'number' && Number.isFinite(v) ? v : fallback;
  };
  const bool = (key: keyof PnHostingPolicy, fallback: boolean): boolean => {
    const v = o[key];
    return typeof v === 'boolean' ? v : fallback;
  };
  const staticPeers = Array.isArray(o.static_peers)
    ? o.static_peers.filter((p): p is string => typeof p === 'string')
    : [];
  const nodeName =
    typeof o.node_name === 'string' && o.node_name.trim() ? o.node_name.trim() : null;
  const coerced: PnHostingPolicy = {
    peering_cost: num('peering_cost', DEFAULT_PN_HOSTING_POLICY.peering_cost),
    max_peering_cost: num('max_peering_cost', DEFAULT_PN_HOSTING_POLICY.max_peering_cost),
    autopeer: bool('autopeer', DEFAULT_PN_HOSTING_POLICY.autopeer),
    autopeer_maxdepth: num('autopeer_maxdepth', DEFAULT_PN_HOSTING_POLICY.autopeer_maxdepth),
    max_peers: num('max_peers', DEFAULT_PN_HOSTING_POLICY.max_peers),
    propagation_stamp_cost: num(
      'propagation_stamp_cost',
      DEFAULT_PN_HOSTING_POLICY.propagation_stamp_cost,
    ),
    propagation_stamp_flex: num(
      'propagation_stamp_flex',
      DEFAULT_PN_HOSTING_POLICY.propagation_stamp_flex,
    ),
    message_storage_limit_mb: num(
      'message_storage_limit_mb',
      DEFAULT_PN_HOSTING_POLICY.message_storage_limit_mb,
    ),
    propagation_limit_kb: num(
      'propagation_limit_kb',
      DEFAULT_PN_HOSTING_POLICY.propagation_limit_kb,
    ),
    sync_limit_kb: num('sync_limit_kb', DEFAULT_PN_HOSTING_POLICY.sync_limit_kb),
    delivery_limit_kb: num('delivery_limit_kb', DEFAULT_PN_HOSTING_POLICY.delivery_limit_kb),
    from_static_only: bool('from_static_only', DEFAULT_PN_HOSTING_POLICY.from_static_only),
    auth_required: bool('auth_required', DEFAULT_PN_HOSTING_POLICY.auth_required),
    enforce_stamps: bool('enforce_stamps', DEFAULT_PN_HOSTING_POLICY.enforce_stamps),
    enforce_ratchets: bool('enforce_ratchets', DEFAULT_PN_HOSTING_POLICY.enforce_ratchets),
    static_peers: staticPeers,
    node_name: nodeName,
    pn_announce_interval_sec: num(
      'pn_announce_interval_sec',
      DEFAULT_PN_HOSTING_POLICY.pn_announce_interval_sec,
    ),
    announce_at_start: bool('announce_at_start', DEFAULT_PN_HOSTING_POLICY.announce_at_start),
  };
  const sanitized = sanitizePnHostingPolicy(coerced);
  if (!sanitized.ok) {
    console.debug(
      '[pnHostingPolicy] invalid policy from sidecar; using defaults: ' + sanitized.error,
    );
    return { ...DEFAULT_PN_HOSTING_POLICY };
  }
  return sanitized.policy;
}

File: reticulum-sidecar/src/stack/pn_hosting_policy.rs
//! Persisted LXMF propagation-node hosting / peering policy.

use serde::{Deserialize, Serialize};

/// Defaults match rsLXMF `RouterConfig` / `RouterConfigExt` / lxmd `[propagation]`.
pub const DEFAULT_PEERING_COST: u8 = 18;
pub const DEFAULT_MAX_PEERING_COST: u8 = 26;
pub const DEFAULT_AUTOPEER: bool = true;
pub const DEFAULT_AUTOPEER_MAXDEPTH: usize = 4;
pub const DEFAULT_MAX_PEERS: usize = 20;
pub const DEFAULT_PROPAGATION_STAMP_COST: u8 = 16;
pub const DEFAULT_PROPAGATION_STAMP_FLEX: u8 = 3;
pub const DEFAULT_MESSAGE_STORAGE_LIMIT_MB: u32 = 256;
pub const DEFAULT_PROPAGATION_LIMIT_KB: usize = 256;
pub const DEFAULT_SYNC_LIMIT_KB: usize = 10_240;
pub const DEFAULT_DELIVERY_LIMIT_KB: usize = 1000;
pub const DEFAULT_PN_ANNOUNCE_INTERVAL_SEC: u32 = 360;
pub const DEFAULT_ANNOUNCE_AT_START: bool = true;

const MAX_AUTOPEER_MAXDEPTH: usize = 64;
const MAX_MAX_PEERS: usize = 256;
/// Cap static peer list size (mirrors TS `MAX_STATIC_PEERS`).
pub const MAX_STATIC_PEERS: usize = 256;
const MAX_STORAGE_MB: u32 = 10_240;
const MAX_LIMIT_KB: usize = 102_400;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
#[allow(clippy::struct_excessive_bools)] // mirrors independent LXMF router prefs
pub struct PnHostingPolicy {
    pub peering_cost: u8,
    pub max_peering_cost: u8,
    pub autopeer: bool,
    pub autopeer_maxdepth: usize,
    pub max_peers: usize,
    pub propagation_stamp_cost: u8,
    pub propagation_stamp_flex: u8,
    pub message_storage_limit_mb: u32,
    pub propagation_limit_kb: usize,
    pub sync_limit_kb: usize,
    pub delivery_limit_kb: usize,
    pub from_static_only: bool,
    pub auth_required: bool,
    pub enforce_stamps: bool,
    pub enforce_ratchets: bool,
    pub static_peers: Vec<String>,
    pub node_name: Option<String>,
    pub pn_announce_interval_sec: u32,
    pub announce_at_start: bool,
}

impl Default for PnHostingPolicy {
    fn default() -> Self {
        Self {
            peering_cost: DEFAULT_PEERING_COST,
            max_peering_cost: DEFAULT_MAX_PEERING_COST,
            autopeer: DEFAULT_AUTOPEER,
            autopeer_maxdepth: DEFAULT_AUTOPEER_MAXDEPTH,
            max_peers: DEFAULT_MAX_PEERS,
            propagation_stamp_cost: DEFAULT_PROPAGATION_STAMP_COST,
            propagation_stamp_flex: DEFAULT_PROPAGATION_STAMP_FLEX,
            message_storage_limit_mb: DEFAULT_MESSAGE_STORAGE_LIMIT_MB,
            propagation_limit_kb: DEFAULT_PROPAGATION_LIMIT_KB,
            sync_limit_kb: DEFAULT_SYNC_LIMIT_KB,
            delivery_limit_kb: DEFAULT_DELIVERY_LIMIT_KB,
            from_static_only: false,
            auth_required: false,
            enforce_stamps: false,
            enforce_ratchets: false,
            static_peers: Vec::new(),
            node_name: None,
            pn_announce_interval_sec: DEFAULT_PN_ANNOUNCE_INTERVAL_SEC,
            announce_at_start: DEFAULT_ANNOUNCE_AT_START,
        }
    }
}

impl PnHostingPolicy {
    pub fn validate(&self) -> Result<(), String> {
        if self.peering_cost > self.max_peering_cost {
            return Err("peering_cost_exceeds_max".into());
        }
        if self.propagation_stamp_flex > self.propagation_stamp_cost {
            return Err("stamp_flex_exceeds_cost".into());
        }
        if self.autopeer_maxdepth > MAX_AUTOPEER_MAXDEPTH {
            return Err("autopeer_maxdepth_out_of_range".into());
        }
        if self.max_peers == 0 || self.max_peers > MAX_MAX_PEERS {
            return Err("max_peers_out_of_range".into());
        }
        if self.message_storage_limit_mb == 0 || self.message_storage_limit_mb > MAX_STORAGE_MB {
            return Err("message_storage_limit_out_of_range".into());
        }
        if self.propagation_limit_kb == 0 || self.propagation_limit_kb > MAX_LIMIT_KB {
            return Err("propagation_limit_out_of_range".into());
        }
        if self.sync_limit_kb == 0 || self.sync_limit_kb > MAX_LIMIT_KB {
            return Err("sync_limit_out_of_range".into());
        }
        if self.delivery_limit_kb == 0 || self.delivery_limit_kb > MAX_LIMIT_KB {
            return Err("delivery_limit_out_of_range".into());
        }
        if self.pn_announce_interval_sec > 86_400 {
            return Err("pn_announce_interval_out_of_range".into());
        }
        if self.static_peers.len() > MAX_STATIC_PEERS {
            return Err("static_peers_too_many".into());
        }
        for peer in &self.static_peers {
            validate_static_peer_hash(peer)?;
        }
        if let Some(name) = &self.node_name {
            let trimmed = name.trim();
            if trimmed.chars().any(char::is_control) {
                return Err("node_name_invalid".into());
            }
            if trimmed.chars().count() > 128 {
                return Err("node_name_too_long".into());
            }
        }
        Ok(())
    }

    /// Clamp and normalize; returns a validated policy or an error for semantic violations.
    pub fn sanitized(mut self) -> Result<Self, String> {
        self.static_peers = self
            .static_peers
            .into_iter()
            .map(|s| s.trim().to_lowercase())
            .filter(|s| !s.is_empty())
            .collect();
        if let Some(name) = self.node_name.take() {
            let trimmed = name.trim().to_string();
            self.node_name = if trimmed.is_empty() {
                None
            } else {
                Some(trimmed)
            };
        }
        self.validate()?;
        Ok(self)
    }

    pub fn message_storage_limit_bytes(&self) -> usize {
        (self.message_storage_limit_mb as usize).saturating_mul(1024 * 1024)
    }

    pub fn min_stamp_cost(&self) -> u8 {
        self.propagation_stamp_cost
            .saturating_sub(self.propagation_stamp_flex)
    }
}

fn validate_static_peer_hash(hash: &str) -> Result<(), String> {
    let trimmed = hash.trim().to_lowercase();
    if trimmed.len() != 32 || !trimmed.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(format!("static_peer_invalid:{trimmed}"));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_validate() {
        assert!(PnHostingPolicy::default().validate().is_ok());
    }

    #[test]
    fn rejects_peering_cost_above_max() {
        let policy = PnHostingPolicy {
            peering_cost: 30,
            max_peering_cost: 26,
            ..Default::default()
        };
        assert_eq!(policy.validate().unwrap_err(), "peering_cost_exceeds_max");
    }

    #[test]
    fn rejects_bad_static_peer() {
        let policy = PnHostingPolicy {
            static_peers: vec!["abcd".into()],
            ..Default::default()
        };
        assert!(
            policy
                .validate()
                .unwrap_err()
                .starts_with("static_peer_invalid:")
        );
    }

    #[test]
    fn rejects_too_many_static_peers() {
        let peer = "aabbccddeeff00112233445566778899".to_string();
        let policy = PnHostingPolicy {
            static_peers: vec![peer; MAX_STATIC_PEERS + 1],
            ..Default::default()
        };
        assert_eq!(policy.validate().unwrap_err(), "static_peers_too_many");
    }

    #[test]
    fn serde_round_trip_defaults() {
        let json = serde_json::to_string(&PnHostingPolicy::default()).unwrap();
        let parsed: PnHostingPolicy = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, PnHostingPolicy::default());
    }

    #[test]
    fn serde_missing_fields_use_defaults() {
        let parsed: PnHostingPolicy = serde_json::from_str("{}").unwrap();
        assert_eq!(parsed.peering_cost, DEFAULT_PEERING_COST);
        assert_eq!(parsed.max_peering_cost, DEFAULT_MAX_PEERING_COST);
        assert!(parsed.autopeer);
    }
}
